### PR TITLE
Add `--target-dir` flag to build and pack command

### LINKED ISSUE #32295: Ability to specify a custom target directory location in `bal build`
This request came from the nBallerina team

### PR DESCRIPTION
**Pull request by Ibaqu**
## Purpose
Adds support to specifiy a target directory to the build and pack command. All caches and built files will be stored in the custom directory. 

Fixes #32295

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by hevayo — DISMISSED** [review]

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java [hunk 1]
```diff
@@ -272,6 +300,12 @@ public void execute() {
                 CommandUtil.exitError(this.exitWhenFinish);
                 return;
             }
+
+            // If a custom target path is set, we modify the targetDir in the build options with the new custom one
+            if (this.targetDir != null) {
+                buildOptions.setTargetDir(this.targetDir.toString());
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java:306]: We shouldn't introduce a setter to BuildOptions. This breaks the immutable model of Project API. The correct way is to use the BuildOptionsBuilder. Check the `constructBuildOptions` method in BuildCommand

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 2]
```diff
@@ -204,7 +222,16 @@ public void setParentCmdParser(CommandLine parentCmdParser) {
     private void pushPackage(BuildProject project) {
         Path balaFilePath = validateBalaFile(project);
         pushBalaToCustomRepo(balaFilePath);
-        outStream.println("Successfully pushed " + userDir.relativize(balaFilePath)
+
+        Path relativePathToBalaFile;
+
+        try {
+            relativePathToBalaFile = project.sourceRoot().relativize(balaFilePath);
+        } catch (IllegalArgumentException e) {
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:230]: Why do we need to catch `IllegalArgumentException` here?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java [hunk 3]
```diff
@@ -78,6 +80,14 @@ public Path sourceRoot() {
         return this.sourceRoot;
     }
 
+    public Path targetDir() {
+        if (this.buildOptions == null || this.buildOptions.getTargetPath() == null) {
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java:84]: We dont need a null check for buildOptions AFAIU. A project is alway created with buildOptions.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java (hunk 2 above)
Comment (Ibaqu) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:230]: When using relative paths for custom target directories, an IllegalArgumentException is thrown. In those cases we catch the exception and handle that case by simply printing the actual custom path.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java (hunk 1 above)
Comment (Ibaqu) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java]: Removed the setter from BuildOptions. Added the condtion to constructBuildOptions() instead

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java (hunk 2 above)
Comment (azinneera) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:231]: We do not need to relativize all the time. If the user has given the path, then we should print it as it is. Else, if is the default one then we print the relative path which is already being done ATM.

Comment (Ibaqu) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java]: Changed the implementation based on if the custom path is given or not

**Comment by Ibaqu** [comment]
Updated paths

-  bal build

```
Generating executable
        target/bin/customDirTest.jar
```

- bal build --target-dir ./custom 

```
Generating executable
        custom/bin/customDirTest.jar
```

- bal build --target-dir ../custom

```
Generating executable
        ../custom/bin/customDirTest.jar
```
-----------------------------------------------------------------------

- bal build --with-tests --code-coverage

```
Generating Test Report
        /Users/mohamedaquibzulfikar/Desktop/33256/target/report/test_results.json
```

- bal build --target-dir ./custom --with-tests --code-coverage

```
Generating Test Report
        ./custom/report/test_results.json
```

- bal build --target-dir ../custom --with-tests --code-coverage

```
Generating Test Report
        ../custom/report/test_results.json
```

-----------------------------------------------------------------------

- bal pack - 

```
Creating bala
        target/bala/ibaqu-customDirTest-any-0.1.0.bala
```

- bal pack --target-dir ./custom

```
Creating bala
        ./custom/bala/ibaqu-customDirTest-any-0.1.0.bala
```

- bal pack --target-dir ../custom

```
Creating bala
        ../custom/bala/ibaqu-customDirTest-any-0.1.0.bala
```

-----------------------------------------------------------------------

- bal push

```
Successfully pushed target/bala/ibaqu-customDirTest-any-0.1.0.bala to 'local' repository.
```

- bal push --target-dir ./custom

```
Successfully pushed ./custom/bala/ibaqu-customDirTest-any-0.1.0.bala to 'local' repository.
```

- bal push --target-dir ../custom

```
Successfully pushed ../custom/bala/ibaqu-customDirTest-any-0.1.0.bala to 'local' repository.
```

**Comment by Ibaqu** [comment]
Changed the implementation of the `push` command

Now instead of specifying the path of the target directory with the `--target-dir` flag, we simply pass the bala file path as an argument. 

`bal push ./custom/bala/<bala-file>`

The validation works exactly the same as the default bal push command.

**Review by azinneera — APPROVED** [review]
LGTM

**Review by sameerajayasoma — APPROVED** [review]
LGTM

**Pull request merged**